### PR TITLE
⭐️ new OPC UA resource pack

### PR DESCRIPTION
**Pull request by chris-rock**
The new OPC UA resource pack allows you to connect to OPC UA server and to query the tree easily via MQL:

```coffeescript
# gather all available namespaces 
opcua.namespaces { * }
opcua.namespaces: [
  0: {
    id: 0
    name: "http://opcfoundation.org/UA/"
  }
  1: {
    id: 1
    name: "urn:open62541.server.application"
  }
]

# gather root node
cnquery> opcua.root
opcua.root: opcua.node id="i=84" name="Root"


# gather all nodes
cnquery> opcua.nodes { name namespace.name }

# gather node with a specific id
cnquery> opcua.nodes.where (id == "i=2253")
opcua.nodes.where: [
  0: opcua.node id="i=2253" name="Server"
]

# gather details about the server
cnquery> opcua.server { * }
opcua.server: {
  buildInfo: {
    BuildDate: "2023-05-21T21:03:43.817369Z"
    BuildNumber: "May 20 2023 15:51:32"
    ManufacturerName: "open62541"
    ProductName: "open62541 OPC UA Server"
    ProductURI: "http://open62541.org"
    SoftwareVersion: "1.3.5-994-g5d73f0cc5"
  }
  node: opcua.node id="i=2253" name="Server"
  currentTime: 2023-05-22 08:28:30.625932 +0000 UTC
  state: "ServerStateRunning"
  startTime: 2023-05-21 21:03:43.834304 +0000 UTC
}
```

Follow ups that need to be implemented on top of this PR:

- add support for authentication modes
- certificate authentication
- more testing with different server implementations
- support querying multiple services with inventories

### CODE REVIEW

File: apps/cnquery/cmd/builder/parse.go [hunk 1]
```diff
@@ -745,6 +745,21 @@ func ParseTargetAsset(cmd *cobra.Command, args []string, providerType providers.
...
+	case providers.ProviderType_OPCUA:
+		connection.Backend = providerType
+
+		cred := &vault.Credential{
+			Type:     vault.CredentialType_password,
+			Password: password,
+		}
+
+		if x, err := cmd.Flags().GetString("endpoint"); err != nil {
+			log.Fatal().Err(err).Msg("cannot parse --host value")
```
Comment (czunker) [inline apps/cnquery/cmd/builder/parse.go]: Perhaps `--endpoint` is clearer than `--host`.

File: motor/discovery/opcua/resolver.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+func (r *Resolver) AvailableDiscoveryTargets() []string {
+	return []string{common.DiscoveryAuto, common.DiscoveryAll}
+}
+
+func (r *Resolver) Resolve(ctx context.Context, root *asset.Asset, cc *providers.Config, credsResolver vault.Resolver, sfn common.QuerySecretFn, userIdDetectors ...providers.PlatformIdDetector) ([]*asset.Asset, error) {
+	resolved := []*asset.Asset{}
+
+	// Note: we use the resolver instead of the direct ms365_provider.New to resolve credentials properly
```
Comment (czunker) [inline motor/discovery/opcua/resolver.go]: I don't think this comment is relevant here.

File: motor/discovery/opcua/resolver.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	m, err := resolver.NewMotorConnection(ctx, cc, credsResolver)
+	if err != nil {
+		return nil, err
+	}
+	defer m.Close()
+
+	provider, ok := m.Provider.(*opcua_provider.Provider)
+	if !ok {
+		return nil, errors.New("could not create ms okta provider")
```
Comment (czunker) [inline motor/discovery/opcua/resolver.go]: "OPC UA" instead of "ms okta"

File: motor/discovery/opcua/resolver.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	// detect platform info for the asset
+	pf, err := m.Platform()
+	if err != nil {
+		return nil, err
+	}
+
+	if cc.IncludesOneOfDiscoveryTarget(common.DiscoveryAuto, common.DiscoveryAll) {
+		resolved = append(resolved, &asset.Asset{
+			PlatformIds: []string{identifier},
+			Name:        "OPC-UA server",
```
Comment (czunker) [inline motor/discovery/opcua/resolver.go]: Everywhere else you use "OPC UA" without the "-". Should this also be the case here?

File: resources/packs/opcua/README.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  state: "ServerStateRunning"
+  startTime: 2023-05-21 21:03:43.834304 +0000 UTC
+}
+```
+
+## Example Servers
+
+*Open62541*
+
+The [Open62541](https://github.com/open62541/open62541) includes may examples for building an Open62541 server.
```
Comment (czunker) [inline resources/packs/opcua/README.md:53]:
```suggestion
The [Open62541](https://github.com/open62541/open62541) includes many examples for building an Open62541 server.
```

File: resources/packs/opcua/opcua.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+var Registry = info.Registry
+
+func init() {
+	Init(Registry)
+}
+
+func opcuaProvider(p providers.Instance) (*opcua_provider.Provider, error) {
+	at, ok := p.(*opcua_provider.Provider)
+	if !ok {
+		return nil, errors.New("okta resource is not supported on this provider")
```
Comment (czunker) [inline resources/packs/opcua/opcua.go]: "okta" looks wrong at this point.

File: resources/packs/opcua/opcua.lr [hunk 7]
```diff
@@ -0,0 +1,65 @@
+option go_package = "go.mondoo.com/cnquery/resources/packs/opcua"
+
+// OPC UA
+opcua {
+  // Namespaces
+  namespaces() []opcua.namespace
+  // Root node
+	root() opcua.node
```
Comment (czunker) [inline resources/packs/opcua/opcua.lr]: The indent changes from here on. Could you please use only one?

File: resources/packs/opcua/opcua.lr [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+	// List of all nodes
+	nodes() []opcua.node
+}
+
+// Server Object
+opcua.server {
+	//
```
Comment (czunker) [inline resources/packs/opcua/opcua.lr]: Are these docs left empty by intent?

**Review by czunker — COMMENTED** [review]
Nice! Thanks @chris-rock 

Only some minor comments.

**Comment by czunker** [comment]
I tried to test this locally, with the Azure container image, but got an error:
```
cnquery shell opcua --endpoint opc.tcp://192.168.1.87:50000 --verbose                                                                                                     
→ loaded configuration from /etc/opt/mondoo/mondoo.yml using source default
DBG parsing asset asset-type=1 provider=OPCUA
→ discover related assets for 1 asset(s)
DBG run resolver resolver="OPC-UA Resolver" resolver-id=opc-ua
DBG establish motor connection
→ resolved assets resolved-assets=0
x could not connect to asset error="opcua: could not resolve address f31e50127436:50000" asset=
FTL could not resolve assets
```

To run the container, I used the command from the README in this PR. Any ideas?

**Comment by chris-rock** [comment]
@czunker Yes, the problem is that the azure opc service returns a resolved url. This essentially highlights that the OPC UA endpoint discovery cannot be trusted for the initial connection. I adjust the code.

**Comment by czunker** [comment]
> @czunker Yes, the problem is that the azure opc service returns a resolved url. This essentially highlights that the OPC UA endpoint discovery cannot be trusted for the initial connection. I adjust the code.

Works now:
```
cnquery> opcua.server { * }
opcua.server: {
  buildInfo: {
    BuildDate: "2023-04-21T11:03:38Z"
    BuildNumber: "1.4.371.86"
    ManufacturerName: "Microsoft"
    ProductName: "IoTEdge OPC UA PLC"
    ProductURI: "https://github.com/Azure/iot-edge-opc-plc.git"
    SoftwareVersion: "1.4.371.86+36dd5653af33ed6af874ef1334d1a98ac28295f3"
  }
  node: opcua.node id="i=2253" name="Server"
  currentTime: 2023-05-30 04:20:25.4249191 +0000 UTC
  state: "ServerStateRunning"
  startTime: 2023-05-30 04:17:20.8710742 +0000 UTC
}
```

Thanks @chris-rock

**Review by czunker — APPROVED** [review]

**Pull request merged**